x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+				return nil, types.ErrLimitationExceed.Wrapf("The lvg number within the bucket exceeds the limit")
+			}
+			// not exist
+			lvgID = k.GenNextLVGID(ctx)
+			lvg = &types.LocalVirtualGroup{
+				Id:                   lvgID,
+				GlobalVirtualGroupId: gvgID,
+				StoredSize:           payloadSize,
+				BucketId:             bucketID,
+			}
+			newLVG = true
+			gvgsBindingOnBucket.AppendGVGAndLVG(gvgID, lvgID)
+		} else {
+			lvg, found = k.GetLVG(ctx, bucketID, lvgID)
+			if !found {
+				return nil, types.ErrLVGNotExist
+			}
+			lvg.StoredSize += payloadSize
+		}
+	}
+
+	gvg.StoredSize += payloadSize
+
+	k.SetGVG(ctx, gvg)
+	k.SetLVG(ctx, lvg)
+	k.SetGVGsBindingOnBucket(ctx, gvgsBindingOnBucket)
+
+	if newLVG {
+		if err := ctx.EventManager().EmitTypedEvents(&types.EventCreateLocalVirtualGroup{
+			Id:                   lvg.Id,
+			BucketId:             lvg.BucketId,
+			GlobalVirtualGroupId: lvg.GlobalVirtualGroupId,
+			StoredSize:           lvg.StoredSize,
+		}); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateLocalVirtualGroup{
+			Id:                   lvg.Id,
+			GlobalVirtualGroupId: lvg.GlobalVirtualGroupId,
+			StoredSize:           lvg.StoredSize,
+		}); err != nil {
+			return nil, err
+		}
+	}
+	return lvg, nil
+}
+
+func (k Keeper) UnBindingObjectFromLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32, payloadSize uint64) error {
+	lvg, found := k.GetLVG(ctx, bucketID, lvgID)
+	if !found {
+		return types.ErrLVGNotExist
+	}
+	gvgsBindingOnBucket, found := k.GetGVGsBindingOnBucket(ctx, bucketID)
+	if !found {
+		panic(fmt.Sprintf("gvgs binding on bucket mapping not found, bucketID: %s", bucketID.String()))
+	}
+	gvgID := gvgsBindingOnBucket.GetGVGIDByLVGID(lvgID)
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		ctx.Logger().Error("GVG Not Exist, bucketID: %s, gvgID: %d, lvgID :%d", bucketID.String(), gvgID, lvgID)
+		return types.ErrGVGNotExist
+	}
+
+	lvg.StoredSize -= payloadSize
+	gvg.StoredSize -= payloadSize
+
+	k.SetLVG(ctx, lvg)
+	k.SetGVG(ctx, gvg)
+	return nil
+}
+
+func (k Keeper) UnBindingBucketFromGVG(ctx sdk.Context, bucketID math.Uint) error {
+	store := ctx.KVStore(k.storeKey)
+
+	gvgsBindingOnBucket, found := k.GetGVGsBindingOnBucket(ctx, bucketID)
+	if !found {
+		return nil
+	}
+
+	for _, lvgID := range gvgsBindingOnBucket.LocalVirtualGroupIds {
+		store.Delete(types.GetLVGKey(bucketID, lvgID))
+	}
+
+	store.Delete(types.GetGVGsBindingOnBucketKey(bucketID))
+	return nil
+}
+
+func (k Keeper) BindingEmptyObjectToGVG(ctx sdk.Context, bucketID math.Uint, primarySPID, familyID uint32) (*types.LocalVirtualGroup, error) {
+	family, found := k.GetGVGFamily(ctx, primarySPID, familyID)
+	if !found {
+		return nil, types.ErrGVGFamilyNotExist
+	}
+
+	if len(family.GlobalVirtualGroupIds) == 0 {
+		return nil, types.ErrGVGNotExist
+	}
+
+	gvgID := family.GlobalVirtualGroupIds[0]
```

Comment:
why pick the first one? Is not gvg specified by SP?

Reply:
Because it's an empty object.It doesn't need sp to sealObject, therefore does not require sp to specify gvg